cmd/events/debug.go

```diff
@@ -0,0 +1,157 @@
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+func newDebugEventsCommand(vp *viper.Viper) *cobra.Command {
+	return &cobra.Command{
+		Use:   "debug",
+		Short: "Get Cilium debug events",
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			debug := vp.GetBool(config.KeyDebug)
+			if err := handleArgs(debug); err != nil {
+				return err
+			}
+			req, err := getDebugEventsRequest()
+			if err != nil {
+				return err
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			hubbleConn, err := conn.New(ctx, vp.GetString(config.KeyServer), vp.GetDuration(config.KeyTimeout))
+			if err != nil {
+				return err
+			}
+			defer hubbleConn.Close()
+			client := observerpb.NewObserverClient(hubbleConn)
+			logger.Logger.WithField("request", req).Debug("Sending GetDebugEvents request")
+			if err := getDebugEvents(client, req); err != nil {
+				msg := err.Error()
+				// extract custom error message from failed grpc call
+				if s, ok := status.FromError(err); ok && s.Code() == codes.Unknown {
+					msg = s.Message()
+				}
+				return errors.New(msg)
+			}
+			return nil
+		},
+	}
+}
+
+func getDebugEventsRequest() (*observerpb.GetDebugEventsRequest, error) {
+	// convert selectorOpts.since into a param for GetDebugEvents
+	var since, until *timestamppb.Timestamp
+	if selectorOpts.since != "" {
+		st, err := hubtime.FromString(selectorOpts.since)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse the since time: %v", err)
+		}
+
+		since = timestamppb.New(st)
+		if err := since.CheckValid(); err != nil {
+			return nil, fmt.Errorf("failed to convert `since` timestamp to proto: %v", err)
+		}
+		// Set the until field if both --since and --until options are specified and --follow
+		// is not specified. If --since is specified but --until is not, the server sets the
+		// --until option to the current timestamp.
+		if selectorOpts.until != "" && !selectorOpts.follow {
+			ut, err := hubtime.FromString(selectorOpts.until)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse the until time: %v", err)
+			}
+			until = timestamppb.New(ut)
+			if err := until.CheckValid(); err != nil {
+				return nil, fmt.Errorf("failed to convert `until` timestamp to proto: %v", err)
+			}
+		}
+	}
+
+	if since == nil && until == nil {
+		switch {
+		case selectorOpts.all:
+			// all is an alias for last=uint64_max
+			selectorOpts.last = ^uint64(0)
+		case selectorOpts.last == 0:
+			// no specific parameters were provided, just a vanilla `hubble events debug`
+			selectorOpts.last = defaults.EventsPrintCount
+		}
+	}
+
+	return &observerpb.GetDebugEventsRequest{
+		Number: selectorOpts.last,
+		Follow: selectorOpts.follow,
+		Since:  since,
+		Until:  until,
+	}, nil
+}
+func getDebugEvents(client observerpb.ObserverClient, req *observerpb.GetDebugEventsRequest) error {
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
Could we simply the signal handling by using [`signal.NotifyContext`](https://golang.org/pkg/os/signal/#NotifyContext) here? Supporting `os.Kill` on the way:
```suggestion
	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt, os.Kill)
```
With this I think we can get rid of the signal handling code in the `go func()` L127 below.

Reply:
This was copied from https://github.com/cilium/hubble/blob/7dcf99de5d47fe4066a79a751593fd6695775407/cmd/observe/observe.go#L562-L563

I guess we'd want to change it there as well (and same for `cmd/events.getAgentEvents`)? Since I'm not exactly sure what the reason for using this construct rather than `signal.NotifyContext` originally in `cmd/observe.getFlows` was, I'd probably defer this to a follow-up PR.

Reply:
I think the only reason is that `signal.NotifyContext` did not exist prior to Go v1.16 :) We probably want to change this in `cmd/observe` as well. I'm happy to tackle this in a follow-up request.

Reply:
Nice, thanks for the explanation. I should have noticed the 1.16 version on [`signal.NotifyContext`](https://golang.org/pkg/os/signal/#NotifyContext) :smile:

Reply:
See #539 for to use `NotifyContext` in the observe command.

Reply:
Done